Background information:
predict whether a potential promotee at a
checkpoint will be promoted or not after the evaluation process.
Defining the Question:
Train a random forest model.
Metric for success:
accuracy should be at least 0.85
Try n_estimators values from 1 to 10. Pick the option with the best quality for the validation set.


In [2]:
#import library
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression



In [6]:
#import our data and preview the data
hrdata_df = pd.read_csv('hrdataset.csv')
hrdata_df.sample()
hrdata_df.tail(3)


employee_id         department     region   education gender  \
54805        13918          Analytics   region_1  Bachelor's      m   
54806        13614  Sales & Marketing   region_9         NaN      m   
54807        51526                 HR  region_22  Bachelor's      m   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
54805               other                1   27                   5.0   
54806            sourcing                1   29                   1.0   
54807               other                1   27                   1.0   

       length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
54805                  3              1            0                  79   
54806                  2              0            0                  45   
54807                  5              0            0                  49   

       is_promoted  
54805            0  
54806            0  
54807            0

In [7]:
#I can explore the last 3 records of the dataset
hrdata_df.tail(3)


employee_id         department     region   education gender  \
54805        13918          Analytics   region_1  Bachelor's      m   
54806        13614  Sales & Marketing   region_9         NaN      m   
54807        51526                 HR  region_22  Bachelor's      m   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
54805               other                1   27                   5.0   
54806            sourcing                1   29                   1.0   
54807               other                1   27                   1.0   

       length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
54805                  3              1            0                  79   
54806                  2              0            0                  45   
54807                  5              0            0                  49   

       is_promoted  
54805            0  
54806            0  
54807            0

In [8]:
#I can explore the first 5 records of the dataset
hrdata_df.head(5)

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [10]:
#check for missing, duplicates, and columns name
# I can check for missing data in the dataset. 
hrdata_df.isnull().sum() #education = 2409 and previous_year_rating=4124



employee_id                0
department                 0
region                     0
education               2409
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    4124
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
is_promoted                0
dtype: int64

In [11]:
# I can check for duplicates in the dataset. 
hrdata_df.duplicated().any() 


False

In [12]:
# I can check for data type in the dataset.
hrdata_df.dtypes #data type okey for the variable values


employee_id               int64
department               object
region                   object
education                object
gender                   object
recruitment_channel      object
no_of_trainings           int64
age                       int64
previous_year_rating    float64
length_of_service         int64
KPIs_met >80%             int64
awards_won?               int64
avg_training_score        int64
is_promoted               int64
dtype: object

In [14]:
#remove columns with missing data 
clean_hrdata_df = hrdata_df.drop(columns=['employee_id','department','gender','recruitment_channel','previous_year_rating', 'education','region'])
clean_hrdata_df.head()


no_of_trainings  age  length_of_service  KPIs_met >80%  awards_won?  \
0                1   35                  8              1            0   
1                1   30                  4              0            0   
2                1   34                  7              0            0   
3                2   39                 10              0            0   
4                1   45                  2              0            0   

   avg_training_score  is_promoted  
0                  49            0  
1                  60            0  
2                  50            0  
3                  50            0  
4                  73            0

In [15]:
#Explore creating and training a model. Attempt to make predictions using the model.
#since there is no separate test dataset create train and validation dataset
#use train and test split function. 
#when using the valid_df no model achieved accuracy of >=0.85
train_df, valid_df = train_test_split(clean_hrdata_df, test_size=0.25, random_state=1234)
print(train_df.shape)
print(valid_df.shape)


(41106, 7)
(13702, 7)


In [16]:
#Come up with features and target for test and train
features_train = train_df.drop(columns=['is_promoted'])
target_train = train_df['is_promoted']
features_valid = valid_df.drop(columns=['is_promoted'])
target_valid = valid_df['is_promoted']

#create a model for Decision Trees, Random Forest and Logistic Regression
#model for Decision Trees, declare and find the ideal depth for the tree
for d in range(1, 11, 1):
  model = DecisionTreeClassifier(random_state=1234, max_depth=d)
  model.fit(features_train, target_train)  #train the model
  #check for accuracy
  print(f'Decision tree has accuracy of: {model.score(features_valid, target_valid)} for depth of: {d}')

# #declare model for random forest and find the best n_estimator value
# for n in range(1,20,1):
#   forest_model = RandomForestClassifier(random_state=1234, n_estimators=n)
#   forest_model.fit(features_train, target_train)
#   print(f'Random forest has accuracy of: {forest_model.score(features_train, target_train)} for n={n}')

# #declare a model for logistic regression
# log_model = LogisticRegression(random_state=1234, solver='liblinear')
# log_model.fit(features_train, target_train)
# print(f'logistic regression has accuracy of: {log_model.score(features_train, target_train)}')




Decision tree has accuracy of: 0.9216172821485914 for depth of: 1
Decision tree has accuracy of: 0.9216172821485914 for depth of: 2
Decision tree has accuracy of: 0.9216172821485914 for depth of: 3
Decision tree has accuracy of: 0.9241716537731718 for depth of: 4
Decision tree has accuracy of: 0.923952707633922 for depth of: 5
Decision tree has accuracy of: 0.923441833309006 for depth of: 6
Decision tree has accuracy of: 0.9235877974018392 for depth of: 7
Decision tree has accuracy of: 0.9227849948912568 for depth of: 8
Decision tree has accuracy of: 0.9232228871697562 for depth of: 9
Decision tree has accuracy of: 0.9217632462414246 for depth of: 10


###Finding and Recommendation
*   Accuracy of 0.92 has been achieved which is above 0.85 minimum set target. Thus, we can predict whether a potential promotee at a checkpoint will be promoted or not after the evaluation process







